### PR TITLE
IpcClient Implementation

### PR DESCRIPTION
**Pull request by hswick**
I found a neat way to do #8 using no dependencies, just :gen_tcp.

### CODE REVIEW

**Comment by hswick** [comment]
Looks like it isn't finding the IPC socket path in the CircleCI build

File: lib/ethereumex/ipc_server.ex [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  def start_link(state \\ %{}) do
+    GenServer.start_link(__MODULE__, Map.merge(state, %{socket: nil}), name: IpcServer)
+  end
+
+  def post(request) do
+    GenServer.call(IpcServer, {:request, request})
+  end
+
+  def handle_call({:request, request}, _from, %{socket: socket} = state) do
+    :ok = :gen_tcp.send(socket, request)
```
Comment (ayrat555) [inline lib/ethereumex/ipc_server.ex]: Can we somehow handle socket/request errors here? Maybe we can pass those errors to `Ethereumex.IpcClient.post_request/2` and return them to a user. What do you think about it?

**Comment by ayrat555** [comment]
@hswick There is a bounty for this issue on [gitcoin](https://gitcoin.co/issue/exthereum/ethereumex/8/647). You should submit your PR (I don't the details, never used it). I

**Comment by hswick** [comment]
Hi @ayrat555, I made the changes you recommended.

File: lib/ethereumex/client/base_client.ex [hunk 2]
```diff
@@ -0,0 +1,39 @@
+defmodule Ethereumex.Client.BaseClient do
+  @moduledoc false
+
+  defmacro __using__(_opts) do    
+    quote do
+      
+      @spec single_request(map(), []) :: {:ok, any() | [any()]} | error
+      def single_request(payload, opts \\ []) do
+	payload
```
Comment (ayrat555) [inline lib/ethereumex/client/base_client.ex]: something's wrong with formatting

Comment (hswick) [reply lib/ethereumex/client/base_client.ex]: Oops. I always forget to run format

File: lib/ethereumex/config.ex [hunk 3]
```diff
@@ -25,4 +25,25 @@ defmodule Ethereumex.Config do
   def request_timeout do
     Application.get_env(:ethereumex, :request_timeout, 5000)
   end
+
+  def setup_children do
+    setup_children(Application.get_env(:ethereumex, :client, :http))
```
Comment (ayrat555) [inline lib/ethereumex/config.ex]: What do you think about setting up children directly in `Ethereumex` module and using `Config` module only for getting configuration params?

File: lib/ethereumex/ipc_client.ex [hunk 4]
```diff
@@ -0,0 +1,21 @@
+defmodule Ethereumex.IpcClient do
+  use Ethereumex.Client.Macro
+  use Ethereumex.Client.BaseClient
+  # import Ethereumex.Config
```
Comment (ayrat555) [inline lib/ethereumex/ipc_client.ex]: unused import

File: config/config.exs [hunk 5]
```diff
@@ -2,4 +2,8 @@ use Mix.Config
 
 config :ethereumex, request_timeout: 5000
 
+config :ethereumex, :client, :ipc
+
+config :ethereumex, :ipc_path, "/.local/share/io.parity.ethereum/jsonrpc.ipc"
```
Comment (ayrat555) [inline config/config.exs]: I think this value is used for testing?

**Comment by ayrat555** [comment]
@hswick Thanks again. Tomorrow I will test these changes locally, will look into `eth_get_block_by_number` (it's one of the most used API methods). I want this PR to get merged as soon as possible

**Comment by hswick** [comment]
@ayrat555 you are very welcome! Thanks for all of the quick feedback. Excited to use this with ExW3. Best of luck with debugging!

**Comment by ayrat555** [comment]
@hswick Can you add me to your fork? I think I found the solution to the truncated json problem.  I want to push changes to your branch

**Comment by hswick** [comment]
Looks like you got the changes on the fork.

I was just thinking the issue was that the packet being sent back wasnt complete, so I’m glad to see you got that figured out. 

And I also agree with your changes about moving the logic to base client.

**Comment by hswick** [comment]
If compiling manually is too slow. There are precompiled binaries of parity for Linux. I’ve found this works well in the past

**Comment by hswick** [comment]
@ayrat555 looks like changing from Macro to BaseClient broke `macro_test.exs`, even swapping out the module being used doesn't fix it. Which is why I didn't change it before. So I may need your help with that because I don't want to completely bulldoze those tests lol

**Comment by ayrat555** [comment]
ok. I'll look into it

File: .circleci/config.yml [hunk 6]
```diff
@@ -11,7 +11,7 @@ jobs:
       - run: echo > passfile
       - run: ./parity --chain 2>&1 &
```
Comment (ayrat555) [inline .circleci/config.yml]: @hswick  you can start parity with `/parity --jsonrpc-interface all --no-warp`. It starts ipc and http servers
`

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,6 @@
+# 0.3.5
```
Comment (ayrat555) [inline CHANGELOG.md]: It's not breaking changes but it's a pretty big feature. maybe 0.4?

Comment (hswick) [reply CHANGELOG.md]: Whatever you want boss :wink:

Comment (ayrat555) [reply CHANGELOG.md]: Also, can you please add these changes https://github.com/exthereum/ethereumex/pull/42  to CHANGELOG?

**Comment by hswick** [comment]
Good news is I got the build working.

Bad news any version of parity above 1.10 makes the tests fail, but I think that should be created on a separate issue.

File: README.md [hunk 8]
```diff
@@ -37,6 +37,24 @@ config :ethereumex,
   request_timeout: 10_000 # default is 5000 ms
 ```
 
+If you want to use IPC you will need to a few things in your config.
```
Comment (ayrat555) [inline README.md]: need to set?

Comment (hswick) [reply README.md]: yep... good catch

**Comment by ayrat555** [comment]
> Bad news any version of parity above 1.10 makes the tests fail, but I think that should be created on a separate issue.
I ran manually compiled parity from the latest master branch. And it seems to work fine.
@hswick  Can you create a separate issue?

**Comment by hswick** [comment]
> > Bad news any version of parity above 1.10 makes the tests fail, but I think that should be created on a separate issue.
> 
> I ran manually compiled parity from the latest master branch. And it seems to work fine.
> @hswick Can you create a separate issue?

#43

File: CHANGELOG.md (hunk 7 above)
Comment (ayrat555) [reply CHANGELOG.md]: ping ^

**Comment by ayrat555** [comment]
@hswick :fire:  Amazing work!

**Comment by ayrat555** [comment]
https://hex.pm/packages/ethereumex/0.4.0

**Pull request merged**